### PR TITLE
USB OTG full-speed support on the stm32u575/585

### PR DESCRIPTION
**Pull request by FRASTM**
This PR adds the support of the USB  on-the-go full-speed (OTG_FS) peripheral on the stm32U575 and stm32U585
and enables the instance of this peripheral on the disco board b_u585i_iot02a
This target offers a USB type C connector.

Signed-off-by: Francois Ramu francois.ramu@st.com

### CODE REVIEW

**Comment by FRASTM** [comment]
testing  _tests/subsys/usb/device/_ 
```
*** Booting Zephyr OS build v2.7.99-846-g6d7c8ebaf4de  ***
D: lock usb_enable_lock mutex
D: 
D: 
D: Pinctrl signals configuration
D: HAL_PCD_Init
E: PCD_Init failed, 1
D: unlock usb_enable_lock mutex
Failed to enable USB
================================================================
PROJECT EXECUTION SUCCESSFUL
```
testing  _tests/subsys/usb/os_desc/_ 


```
*** Booting Zephyr OS build v2.7.99-846-g6d7c8ebaf4de  ***
Running test suite test_osdesc
================================================================
START - test_osdesc_string
test_register_osdesc
test_handle_os_desc: ret 0 len 18 data 0x200000a8
 PASS - test_osdesc_string in 0.7 seconds
================================================================
START - test_osdesc_feature
test_register_osdesc
test_handle_os_desc_feature: ret 0 len 40 data 0x20000080
 PASS - test_osdesc_feature in 0.8 seconds
================================================================
Test suite test_osdesc succeeded
================================================================
PROJECT EXECUTION SUCCESSFUL
```

**Comment by FRASTM** [comment]
testing  _tests/subsys/usb/desc_sections/_ 
```
*** Booting Zephyr OS build v2.7.99-846-g6d7c8ebaf4de  ***
Running test suite test_desc
================================================================
START - test_desc_sections
__usb_descriptor_start 0x200000c0
__usb_descriptor_end 0x2000015d
USB Descriptor table span 157
__usb_data_start 0x20000160
__usb_data_end 0x200001a8
USB Configuration data span 72
sizeof usb_cfg_data 36
D: Starting logs
D: USB Descriptor table section
D: 12 01 00 02 00 00 00 40 |.......@
D: e3 2f 00 01 07 02 01 02 |./......
D: 03 01 09 02 45 00 02 01 |....E...
D: 00 c0 32 09 04 00 00 03 |..2.....
D: ff 00 00 00 07 05 01 02 |........
D: 40 00 00 07 05 81 02 40 |@......@
D: 00 00 07 05 82 02 40 00 |......@.
D: 00 09 04 01 00 03 ff 00 |........
D: 00 00 07 05 02 02 40 00 |......@.
D: 00 07 05 83 02 40 00 00 |.....@..
D: 07 05 84 02 40 00 00 04 |....@...
D: 03 09 04 0e 03 5a 00 45 |.....Z.E
D: 00 50 00 48 00 59 00 52 |.P.H.Y.R
D: 00 10 03 55 00 53 00 42 |...U.S.B
D: 00 2d 00 44 00 45 00 56 |.-.D.E.V
D: 00 22 03 33 00 35 00 33 |.".3.5.3
D: 00 36 00 35 00 30 00 30 |.6.5.0.0
D: 00 34 00 30 00 30 00 34 |.4.0.0.4
D: 00 45 00 30 00 30 00 32 |.E.0.0.2
D: 00 44 00 00 00          |.D...   
D: USB Configuratio structures section
D: 00 00 00 00 db 00 00 20 |....... 
D: 09 0a 00 08 00 00 00 00 |........
D: 00 00 00 00 00 00 00 00 |........
D: 00 00 00 00 03 00 00 00 |........
D: 28 00 00 20 00 00 00 00 |(.. ....
D: f9 00 00 20 09 0a 00 08 |... ....
D: 00 00 00 00 00 00 00 00 |........
D: 00 00 00 00 00 00 00 00 |........
D: 03 00 00 00 40 00 00 20 |....@.. 
D: iface 0
D: found ep[0] 1
D: found ep[1] 81
D: found ep[2] 82
D: iface 1
D: found ep[0] 2
D: found ep[1] 83
D: found ep[2] 84
 PASS - test_desc_sections in 0.134 seconds
================================================================
Test suite test_desc succeeded
================================================================
PROJECT EXECUTION SUCCESSFUL
```

**Comment by FRASTM** [comment]
testing tests/subsys/usb/bos/
```
*** Booting Zephyr OS build v2.7.99-846-g6d7c8ebaf4de  ***
Running test suite test_osdesc
================================================================
START - test_usb_bos_macros
length 57
D: Header
D: 05 0f 39 00 02 18 10 05 |..9.....
D: 00 38 b6 08 34 a9 09 a0 |.8..4...
D: 47 8b fd a0 76 88 15 b6 |G...v...
D: 65 00 01 01 01 1c 10 05 |e.......
D: 00 df 60 dd d8 89 45 c7 |..`...E.
D: 4c 9c d2 65 9d 9e 64 8a |L..e..d.
D: 9f 00 00 03 06 03 00 02 |........
D: 00                      |.       
D: webusb cap
D: 05 0f 39 00 02 18 10 05 |..9.....
D: 00 38 b6 08 34 a9 09 a0 |.8..4...
D: 47 8b fd a0 76 88 15 b6 |G...v...
D: webusb cap msos v2
D: 05 0f 39 00 02 1c 10 05 |..9.....
D: 00 df 60 dd d8 89 45 c7 |..`...E.
D: 4c 9c d2 65 9d 9e 64 8a |L..e..d.
D: 9f 00 00 03             |....    
 PASS - test_usb_bos_macros in 0.55 seconds
================================================================
START - test_usb_bos
test_usb_bos: ret 0 len 57 data 0x2000013d
 PASS - test_usb_bos in 0.4 seconds
================================================================
Test suite test_osdesc succeeded
================================================================
PROJECT EXECUTION SUCCESSFUL
```

**Comment by FRASTM** [comment]
rebase fix conflit on  boards/arm/b_u585i_iot02a/b_u585i_iot02a.dts

**Review by erwango — CHANGES_REQUESTED** [review]
@FRASTM I suspect `tests/subsys/usb/` test suite only focuses on subsystem and not device driver.
Would you mind testing on `samples/subsystem/usb/cdc_acm` ?
I'm getting following error on my side:
```
*** Booting Zephyr OS build v2.7.99-929-gf5f86318181a  ***                      
[00:00:00.016,000] <err> usb_dc_stm32: PCD_Init failed, 1                       
[00:00:00.016,000] <err> cdc_acm_echo: Failed to enable USB 
```
One thing to check would be the USB specific clock.

**Comment by FRASTM** [comment]
adding the HSI48 clock enable for stm32u5x

**Comment by FRASTM** [comment]
testing samples/subsys/usb/cdc_acm:
```
*** Booting Zephyr OS build v2.7.99-930-g031f409017db  ***            
[00:00:00.006,000] <inf> cdc_acm_echo: Wait for DTR                   
[00:00:00.012,000] <inf> usb_cdc_acm: Device suspended                
[00:00:00.575,000] <inf> usb_cdc_acm: Device configured               
[00:00:01.707,000] <inf> cdc_acm_echo: DTR set                        
[00:00:02.722,000] <inf> cdc_acm_echo: Baudrate detected: 115200   
```

File: boards/arm/b_u585i_iot02a/b_u585i_iot02a.dts [hunk 1]
```diff
@@ -67,6 +67,11 @@
 	status = "okay";
 };
 
+zephyr_udc0: &usbotg_fs {
```
Comment (erwango) [inline boards/arm/b_u585i_iot02a/b_u585i_iot02a.dts:70]: Please update board doc

**Review by erwango — CHANGES_REQUESTED** [review]
Board doc missing otherwise this is fine

**Review by erwango — APPROVED** [review]

**Comment by FRASTM** [comment]
doc of the` boards/arm/b_u585i_iot02a/ ` updated

**Review by MaureenHelm — APPROVED** [review]

**Pull request merged**